### PR TITLE
[addons][audiodecoder] fix conflict problems if Kodi and addon supports same

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

This change deals with the problems that occur with ISOs when using Video ISO and SACD ISO. The associated addon then comes into conflict with Kodi's way.

With this, the add-on is called up first to ensure correct use. If it fails then Kodi's own way of processing is used.

The change is still small at the moment and relates only to audio decoder addons, which can support multiple tracks in one file (similar to an archive file).

## Motivation and context
Relates to this issues:
- https://github.com/xbmc/audiodecoder.sacd/issues/18 (The bad in german :smirk:)
- https://github.com/xbmc/audiodecoder.sacd/issues/14
- https://github.com/xbmc/audiodecoder.sacd/issues/11

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/utils/FileExtensionProvider.cpp [hunk 1]
```diff
@@ -92,6 +94,57 @@ std::string CFileExtensionProvider::GetFileFolderExtensions() const
...
+        {
+          /* Call addon to start a dir read about given file, if success, return
+           * as true.
+           */
+          CAudioDecoder result(addonInfo);
+          if (result.CreateDecoder() && result.ContainsFiles(CURL(path)))
+            return true;
+
+          /* If extension was supported and addon creation failed, we expect
+           * as not usable file and give as false here.
```
Comment (garbear) [inline xbmc/utils/FileExtensionProvider.cpp:126]: Grammar in comments this time looks good. I would change this one to

> If extension is supported and addon creation failed, we expect the file is not usable and return false here.

File: xbmc/utils/FileExtensionProvider.cpp [hunk 2]
```diff
@@ -92,6 +94,57 @@ std::string CFileExtensionProvider::GetFileFolderExtensions() const
...
+          return false;
+        }
+      }
+    }
+
+    /*!
+     * About VFS addons we expect that they can and return true.
+     *
+     * @todo Check VFS addons can also be types in conflict with Kodi's
+     * supported parts!
```
Comment (garbear) [inline xbmc/utils/FileExtensionProvider.cpp:137]: For clarity I would use these words

> We expect that VFS addons can support the file, and return true

Without the comma it reads as "We expect that VFS addons can support the file and we expect that VFS addons return true."

With the comma, it shifts the subject and reads as "We expect that VFS addons can support the file, and we return true."

Comment (AlwinEsch) [reply xbmc/utils/FileExtensionProvider.cpp:137]: Got updated, thanks for the help.

**Comment by AlwinEsch** [comment]
Request OK to come in?

Is initial, will look next time on new "CanOperateExtension" with API changes within VFS and audiodecoder to bring in better ways, but first to have equal to needed backward (where no API change allow) and to become the addon issues closed soon.

I may also combine it with a planned request for a GUI listing in Kodi for all supported file extensions and mime types.
Planned idea:
- New window / dialog which lists all mimetypes and file extensions available for Kodi, with description
   - So that the user can see what Kodi can do
- User selection which is favored for editing (respective addon or Kodi first)
   - Possibly useful with e.g. vfs.libarchive, which can also perform formats that Kodi already has itself
- Option to list all possible and, in the meantime, allow users to install the associated add-on
- Individual types strictly forbidden by the user if desired

**Comment by garbear** [comment]
Extension/mine type issues are also a big problem with game add-ons, given that ISOs can be DVDs or PlayStation discs. Any idea how to fix that?

**Comment by AlwinEsch** [comment]
> Extension/mine type issues are also a big problem with game add-ons, given that ISOs can be DVDs or PlayStation discs. Any idea how to fix that?
> 
> EDIT: See [xbmc/vfs.libarchive#53](https://github.com/xbmc/vfs.libarchive/pull/53)

I thought that problems would arise in other areas as well.

Will then add API functions to all addons which use "mimetypes" and "extensions" for upcoming API changes (after this request is there).

Probably even in CAddonDll to have it type independent and to avoid lively class loading during exams.

**Review by garbear — APPROVED** [review]

**Pull request closed without merging**